### ISSUE TITLE
[rush] `bin` field is ignored for linked packages

### ISSUE THREAD

**Issue by lhorie** [Microsoft/web-build-tools#1100]
If a rush workspace contains two projects, one of which has a `bin` script and the other depends on the first, the second project will not get a `node_modules/.bin` symlink to the script of the first.

See this repo for repro steps: https://github.com/lhorie/rush-bin-bug

PS: I'm willing to submit a PR to fix this if someone can guide me to the appropriate whereabouts

**Comment by octogonz** [comment]
Your repro is using the style where NPM itself generates the binary script during installation.  Since Rush links (rather than installs) the local projects, rush would need to reproduce that binary-generating heuristic, which is nontrivial and a little different for yarn/pnpm/npm.

For our projects, we instead explicitly define the binary script, like this:

https://github.com/Microsoft/web-build-tools/blob/master/apps/api-extractor/bin/api-extractor

The **api-documenter** project depends on **api-extractor**.  If you run `rush install` in the **web-build-tools** repo, you will see that a binary gets written here:

`./apps/api-documenter/node_modules/.bin/api-extractor`

Would that solve your problem?

**Comment by lhorie** [comment]
Sorry, I'm not sure I'm following. What's different between the repro repo and the way api-extractor gets written into the node_modules/.bin of api-documenter in the web-build-tools repo?

**Comment by Toxaris** [comment]
We are having the same issue. Based on the discussion in #915 and #1024, we are guessing that `node_modules/.bin` only work for using rush with pnpm but not for using rush with yarn.

**Comment by octogonz** [comment]
@lhorie sorry I overlooked your reply back in Feb.

The difference is that with api-extractor's, the symlink target exists at the time when `rush install` is running, because the target is a source file tracked by Git.  Whereas with the other approach, `rush install` would need to create a symlink to a potentially nonexistent target (i.e. that won't exist until after `rush build` completes).  This can be problematic, e.g. if I remember right, we create hardlinks on Windows, which require the target to exist.

Does that make sense?

**Comment by octogonz** [comment]
@Toxaris I agree that perhaps all these issues could be solved by updating Yarn/NPM to use the more sophisticated `pnpmLinkBins()` function mentioned in https://github.com/Microsoft/web-build-tools/issues/915#issuecomment-434466948 .  Even though it's part of the PNPM package manager library, if I understand right this function should work fine for Yarn and NPM.

This should be an easy fix.  Maybe someone could create a PR?

**Comment by chrisdothtml** [comment]
Bit late to the party, but after testing with @lhorie 's repro repo, @octogonz 's recommended solution isn't actually why it's working for `api-extractor`. It still doesn't work after converting the `my-cli` to an explicit binary. The only reason `api-extractor` is working is because it's using pnpm instead of yarn.

Looks like the same issue exists with npm, so I think the *only* solution is something like `pnpmLinkBins` for npm and yarn

**Comment by chrisdothtml** [comment]
PR made for this: #1256

**Comment by bweggersen** [comment]
We just hit this issue when trying to add Rush to our repo using yarn as the package manager. Specifically, we can't see any bin scripts added to `node_modules/.bin` from any of the internal dependencies. Here's a repro repo: https://github.com/bweggersen/rush-bin-bug-yarn

**Comment by yeegor** [comment]
Still experiencing issues with this. Please fix! 😁

**Comment by mshimokura** [comment]
We are also still experiencing this issue. A workaround is to `rush install --only <package with bin>` and `rush build --only <package with bin>` the binary first, then it will be linked into `<package requires bin>/node_modules/.bin`
